src/payment/strategies/masterpass/masterpass-payment-strategy.ts

```diff
@@ -0,0 +1,127 @@
+import { PaymentStrategy } from '../';
+import {
+    PaymentActionCreator,
+    PaymentInitializeOptions,
+    PaymentMethodActionCreator,
+    PaymentRequestOptions
+} from '../../';
+import { Checkout, CheckoutStore, InternalCheckoutSelectors } from '../../../checkout';
+import {
+    InvalidArgumentError,
+    MissingDataError,
+    MissingDataErrorType
+} from '../../../common/error/errors';
+import { StoreConfig } from '../../../config';
+import { OrderActionCreator, OrderRequestBody } from '../../../order';
+import { PaymentArgumentInvalidError } from '../../errors';
+import PaymentMethod from '../../payment-method';
+
+import { MasterpassCheckoutCallback, MasterpassCheckoutOptions } from './masterpass';
+import MasterpassScriptLoader from './masterpass-script-loader';
+
+export default class MasterpassPaymentStrategy extends PaymentStrategy {
+    private _paymentMethod?: PaymentMethod;
+    private _paymentGateway?: string;
+
+    constructor(
+        store: CheckoutStore,
+        private _orderActionCreator: OrderActionCreator,
+        private _paymentActionCreator: PaymentActionCreator,
+        private _paymentMethodActionCreator: PaymentMethodActionCreator,
+        private _masterpassScriptLoader: MasterpassScriptLoader
+    ) {
+        super(store);
+    }
+
+    initialize(options: PaymentInitializeOptions): Promise<InternalCheckoutSelectors> {
+        const { methodId, masterpass: masterpassOptions } = options;
+
+        if (!masterpassOptions) {
+            throw new InvalidArgumentError('Unable to initialize payment because masterpass options is missing');
+        }
+
+        const state = this._store.getState();
+        this._paymentMethod = state.paymentMethods.getPaymentMethod(methodId);
+
+        return Promise.resolve(this._hasPaymentInfo()).then(hasPaymentInfo => {
+            if (!hasPaymentInfo) {
+                return this._masterpassClientSetup().then(checkoutCallback => checkoutCallback());
```

Comment:
Sorry, I don't 100% follow here. You want to launch the Masterpass modal (or redirect to Masterpass website) when `initialize` is called? This can't happen because `initialize` is suppose to be a preparation step, i.e.: loading some scripts, attaching some event listeners. Nothing should actually happen until `execute` is called, or trigger any other event listeners themselves.

Reply:
If your intention is to prompt the shopper if they haven't selected a payment option previously, you need to trigger the modal / perform the redirect in the `execute` step. You'll know what I mean if you take a look at the implementation of Masterpass provided by Square.

Reply:
Yes when the masterpass payment provider (wallet) is selected on the payment form the redirect is made, the following link contains a video showing that functionality https://drive.google.com/open?id=1NWpm69noGMhjsg09kB1-Vb5v3H4a_G-1